models/address_book.rb

```diff
@@ -31,25 +33,25 @@ def import_from_csv(file_name)
 
     csv.each do |row|
       row_hash = row.to_hash
-      add_entry(row_hash['name'], row_hash['phone_number'], row_hash['email'])
+      add_entry(row_hash["name"], row_hash["phone_number"], row_hash["email"])
     end
   end
 
-  def binary_search(name)
-    lower = 0
-    upper = entries.length - 1
-
-    while lower <= upper
-      mid = (lower + upper) / 2
-      mid_name = entries[mid].name
+  def binary_search(name, lower, upper)
+    mid = (upper + lower) / 2
+    
+    if lower >= upper || entries[mid].name == name
+      entries[mid].name == name ? entries[mid] : nil
```

Comment:
It looks like you are checking the same condition twice. Maybe you can refactor it to have use a "guard clause" instead :smiley: :

```ruby
mid = (upper + lower) / 2
return entries [mid] if entries[mid].name == name

# Rest of the code
...
```